### ISSUE TITLE
Providing Config.__getitem__() is error prone

### ISSUE THREAD

**Issue by tartley** [canonical-ols/talisker#567]
In more than one of the services which use talisker, there have been bugs caused by callers which should have used:

```
talisker.get_config()['DEVEL']
# or
talisker.get_config().devel
```

but instead used:

```
talisker.get_config()['devel']
```

This formulation doesn't work, it just returns None.

See definition of __getitem__() at
https://github.com/canonical-ols/talisker/blob/8864e3710d08fa8c44e3d749f3a355345ab52864/talisker/config.py#L148

The above makes me wonder whether offering `__getitem__` on the Config object itself is wise, or does it just potentially confuse callers. Perhaps it would be nice to offer it as a function like `Config.raw(key)`, so it's more explicit whether callers should be passing nice property names like 'devel', or raw config keys like 'DEVEL'.

Similarly, should we fall back to returning None for unrecognized keys, or raise an index error.

 Of course these are compatibility breaking, so would require care to introduce, but perhaps they would save future bugs like the ones we've already had.

**Comment by maxiberta** [comment]
I think raising a KeyError on a missing key makes sense and is the expected behaviour from a dict-like object, while keeping full backwards compatibility.

**Comment by tartley** [comment]
@maxiberta Thanks. Yes, that wouldn't be an incompatible API change, but obviously it would still need deploying with care - there may be code in services which is currently passing an invalid key, getting None back and happily carrying on. After the change it will KeyError.

**Comment by maxiberta** [comment]
Having services visibly blow up is good actually, and much better than subtly returning unexpected values.

**Comment by tartley** [comment]
Yes, absolutely agreed, that's why I suggest it! I'm only saying that when we deploy services using this new version of talisker, we should take care that we're not deploying something that would blow up *immediately*, is all.
